rudder-core/src/main/scala/com/normation/rudder/repository/jdbc/ComplianceRepository.scala

```diff
@@ -0,0 +1,133 @@
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.repository.jdbc
+
+
+import net.liftweb.common.Box
+import com.normation.rudder.domain.reports.NodeStatusReport
+import com.normation.rudder.repository.ComplianceRepository
+import com.normation.rudder.db.Doobie
+import com.normation.rudder.db.Doobie._
+import scalaz.{Failure => _, _}, Scalaz._
+import doobie.imports._
+import scalaz.concurrent.Task
+import doobie.contrib.postgresql.pgtypes._
+import com.normation.rudder.db.DB
+import com.normation.rudder.services.reports._
+import com.normation.inventory.domain.NodeId
+import org.joda.time.DateTime
+import com.normation.rudder.domain.reports.RunComplianceInfo
+import com.normation.rudder.domain.reports.ComplianceLevel
+import com.normation.rudder.domain.reports.AggregatedStatusReport
+import com.normation.rudder.domain.reports.CompliancePercent
+
+
+final case class RunCompliance(
+    nodeId      : NodeId
+  , runTimestamp: DateTime
+  , endOfLife   : DateTime
+  , runInfo     : (RunAndConfigInfo, RunComplianceInfo)
+  , summary     : CompliancePercent
+  , details     : AggregatedStatusReport
+)
+
+object RunCompliance {
+
+  def from(runTimestamp: DateTime, endOfLife: DateTime, report: NodeStatusReport) = {
+    RunCompliance(report.nodeId, runTimestamp, endOfLife, (report.runInfo, report.statusInfo), report.compliance.pc, report.report)
+  }
+
+}
+
+class ComplianceJdbcRepository(doobie: Doobie) extends ComplianceRepository {
+  import doobie._
+
+  /*
+   * Save a list of node compliance reports
+   */
+  override def saveRunCompliance(reports: List[NodeStatusReport]): Box[List[NodeStatusReport]] = {
+    /*
+     * some sorting of things. We must only store information about node status reports with
+     * a run
+     */
+    val runCompliances = reports.flatMap { r => r.runInfo match {
+      //ignore case with no runs
+      case _:NoReportInInterval |
+           NoRunNoExpectedReport |
+           _:ReportsDisabledInInterval => None
+
+      case x:Pending => x.optLastRun match {
+        case None =>
+          None
+        case Some((runTime, expected)) =>
+          Some(RunCompliance.from(runTime, x.expirationDateTime, r))
+      }
+
+      case x:NoExpectedReport =>
+        // here, the expiration date has not much meaning, since we don't have
+        // information on that node configuration (and so the node has most likelly no
+        // idea whatsoever of any config, even global). Take default values,
+        // ie 5min for run + 5min for grace
+        Some(RunCompliance.from(x.lastRunDateTime, x.lastRunDateTime.plusMinutes(10), r))
```

Comment:
Today, we do compule compliance on that case. Not sure how (un)successful it is, but we need to just be exaustive on the error cases we can encounter.